### PR TITLE
Add TimeZoneInfo_DisplayNameStartsWithOffset test

### PR DESCRIPTION
**Pull request by EgorBo**
Tests https://github.com/dotnet/coreclr/pull/20728
/cc @tarekgh

### CODE REVIEW

**Comment by tarekgh** [comment]
looks we have a build error 

```
[D:\j\workspace\windows-TGrou---2a8f9c29\src\tests.builds]
15:27:14 System\TimeZoneInfoTests.cs(2224,37): error CS0103: The name 'Regex' does not exist in the current context [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Runtime\tests\System.Runtime.Tests.csproj]
```

**Comment by stephentoub** [comment]
> maybe you need to reference System.Text.RegularExpressions?

It likely needs to import the right namespace, e.g. `using System.Text.RegularExpressions;`.

**Comment by danmosemsft** [comment]
Test failures --
https://mc.dot.net/#/user/EgorBo/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/5d9b296dcc652ffe8a08f44eb7f36689d8672968/workItem/System.Runtime.Tests/analysis/xunit/System.Tests.TimeZoneInfoTests~2FTimeZoneInfo_DisplayNameStartsWithOffset

I guess it just didn't pick up the CoreCLR change yet but that might not explain why one of the failures is Nano-only. Maybe worth logging which one it's failing on, and skipping it for Nano.

File: src/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 1]
```diff
@@ -2211,6 +2212,21 @@ public static void TimeZoneInfo_DaylightDeltaIsNoMoreThan12Hours()
...
+        public static void TimeZoneInfo_DisplayNameStartsWithOffset()
+        {
+            foreach (TimeZoneInfo tzi in TimeZoneInfo.GetSystemTimeZones())
+            {
+                if (tzi.Id != "UTC")
+                {
+                    Assert.False(string.IsNullOrWhiteSpace(tzi.StandardName));
+                    Assert.Matches(@"^\(UTC(\+|-)[0-9]{2}:[0-9]{2}\) \S.*", tzi.DisplayName);
+                    string offset = Regex.Match(tzi.DisplayName, @"(-|)[0-9]{2}:[0-9]{2}").Value;
+                    Assert.Equal(tzi.BaseUtcOffset, TimeSpan.Parse(offset));
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/System.Runtime/tests/System/TimeZoneInfoTests.cs]: yeah should be ok but this assert was meant to check just that it's a timespan there, not a some random string :-)

File: src/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 2]
```diff
@@ -2211,6 +2212,21 @@ public static void TimeZoneInfo_DaylightDeltaIsNoMoreThan12Hours()
...
 
+        [Fact]
+        public static void TimeZoneInfo_DisplayNameStartsWithOffset()
+        {
+            foreach (TimeZoneInfo tzi in TimeZoneInfo.GetSystemTimeZones())
+            {
+                if (tzi.Id != "UTC")
+                {
+                    Assert.False(string.IsNullOrWhiteSpace(tzi.StandardName));
+                    Assert.Matches(@"^\(UTC(\+|-)[0-9]{2}:[0-9]{2}\) \S.*", tzi.DisplayName);
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/System.Runtime/tests/System/TimeZoneInfoTests.cs]: yeah tested with:
```csharp
var span = TimeSpan.FromHours(1.5f);
var items = CultureInfo.GetCultures(CultureTypes.AllCultures)
    .Where(c => span.ToString("hh\\:mm", c) != "01:30")
    .ToArray(); // empty
```

File: src/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 3]
```diff
@@ -2211,6 +2212,21 @@ public static void TimeZoneInfo_DaylightDeltaIsNoMoreThan12Hours()
...
         }
 
+        [Fact]
+        public static void TimeZoneInfo_DisplayNameStartsWithOffset()
+        {
+            foreach (TimeZoneInfo tzi in TimeZoneInfo.GetSystemTimeZones())
+            {
+                if (tzi.Id != "UTC")
+                {
+                    Assert.False(string.IsNullOrWhiteSpace(tzi.StandardName));
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/System.Runtime/tests/System/TimeZoneInfoTests.cs]: @tarekgh I wanted to add a similar check for DisplayName but it's basically also tested with the following Regex for it.

File: src/System.Runtime/tests/System/TimeZoneInfoTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/System.Runtime/tests/System/TimeZoneInfoTests.cs]: What I meant here is, when calling DisplayName on Windows, we get the localized resource for the display name from Windows OS DLL  tzres.dll.mui. for example if you are running on Chinese localized Windows, the display name will come from tzres.dll.mui with the Chinese localized resources. your code you used to try with different cultures is not really affecting anything for how we get the localized resources. 

Anyway, as I mentioned, I did myself a sanity check and I am expecting your test will work fine when running on localized Windows OS.

**Comment by tarekgh** [comment]
@dotnet-bot test this please

**Comment by tarekgh** [comment]
we should wait getting the updated coreclr to corefx for merge these tests.

**Comment by tarekgh** [comment]
@dotnet-bot test this please

**Comment by tarekgh** [comment]
test Windows x64 Debug Build

**Comment by tarekgh** [comment]
@EgorBo 

Could you please have a look at the 2 failing cases? 

on OSX: 
```
Assert.Matches() Failure:
Regex: ^\\(UTC(\\+|-)[0-9]{2}:[0-9]{2}\\) \\S.*
Value: GMT-11:00
Stack Trace :
   at System.Tests.TimeZoneInfoTests.TimeZoneInfo_DisplayNameStartsWithOffset() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Runtime/tests/System/TimeZoneInfoTests.cs:line 2223
```

on Linux: 

```
Assert.Matches() Failure:
Regex: ^\\(UTC(\\+|-)[0-9]{2}:[0-9]{2}\\) \\S.*
Value: Niue Time
Stack Trace :
   at System.Tests.TimeZoneInfoTests.TimeZoneInfo_DisplayNameStartsWithOffset() in /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/System.Runtime/tests/System/TimeZoneInfoTests.cs:line 2224
```

**Comment by tarekgh** [comment]
test Windows x64 Debug Build

**Comment by tarekgh** [comment]
```
Windows.10.Nano.Amd64.Open-x64-Debug
Get Repro environment
Unhandled Exception of Type Xunit.Sdk.EqualException
Message :
Assert.Equal() Failure
Expected: -04:00:00
Actual:   -05:00:00
Stack Trace :
   at System.Tests.TimeZoneInfoTests.TimeZoneInfo_DisplayNameStartsWithOffset() in D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime\tests\System\TimeZoneInfoTests.cs:line 2225

```

**Comment by EgorBo** [comment]
@tarekgh 
the problematic timezone is `(UTC-05:00) Turks and Caicos`
so for some reason Display name differs from what actually is in `BaseOffset` property (which is -04:00:00).
So it looks more like a bug actually or maybe the CI machine needs to be restarted?

**Comment by tarekgh** [comment]
This is a Windows Nano server bug which has the wrong TZ data in the registry for Turks and Caicos. 

could you please modify your test to exclude this case on Nano server or at least allow both values (-4:00 and -5:00)? you can use 

https://github.com/dotnet/corefx/blob/master/src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.cs#L31

to check if you are running on Nano

**Comment by tarekgh** [comment]
thanks @EgorBo

**Pull request merged**